V0.29: updates to auto-batching (#1864)
meilisearch/documentation: learn/core_concepts/documents.md

::: note
If you don't specify the data type for an attribute, it will default to `:string`.
:::

### Auto-batching

> Auto-batching combines consecutive document addition requests into a batch to be processed together. This significantly speeds up the indexing process.

Reviewer: If you do anything about my comment on `quick_start.md`, you should probably also change this line.

> For document addition requests to be added to the same batch, they need to:

Reviewer: Purely aesthetic, not at all a required change.

```suggestion
Meilisearch batches document addition requests when they:
```

- Target the same index
- Have the same update method (i.e., [POST](/reference/api/documents.md#add-or-replace-documents) or [PUT](/reference/api/documents.md#add-or-update-documents))

> - Be immediately consecutive

Reviewer: If you accept my suggestion above, you should also change this list item.

> Tasks within the same batch share the same values for `startedAt`, `finishedAt`, `duration`, and the same `error` object, if an [internal](/reference/api/overview.md#errors) error occurs. If a task fails due to an invalid document, it will not be processed with the batch and will have its own error message.

Reviewer 2: I feel that some clarification is required here on what will happen if a task fails due to an invalid document. Will all the other tasks that were part of the same batch still complete? I think yes, so I'm suggesting this change:

```suggestion
Tasks within the same batch share the same values for `startedAt`, `finishedAt`, `duration`, and the same `error` object, if an [internal](/reference/api/overview.md#errors) error occurs. If a task fails due to an invalid document, it will be removed from the batch. The rest of the batch will still process normally.
```

> Tasks within the same batch share the same values for `startedAt`, `finishedAt`, `duration`, and the same `error` object, if an [internal](/reference/api/overview.md#errors) error occurs. If a task fails due to an invalid document, it will be removed from the batch. The rest of the batch will still process normally.

Reviewer: ```suggestion
Tasks within the same batch share the same values for `startedAt`, `finishedAt`, and `duration`. They also share the same `error` object, if an [internal](/reference/api/overview.md#errors) error occurs. If a task fails due to an invalid document, it will be removed from the batch. The rest of the batch will still process normally.
```

> You can deactivate auto-batching using the `--disable-auto-batching` command-line flag or the `MEILI_DISABLE_AUTO_BATCHING` environment variable.

Reviewer: Perhaps we can describe a situation where disabling auto-batching might be useful? Just one short sentence.

I feel it might be necessary in this case because it's unclear to me what I, a Meilisearch user, could possibly gain from disabling this feature.

Author: [Added](https://github.com/meilisearch/documentation/issues/1838#issuecomment-1248011040)

::: note
If you don't specify the data type for an attribute, it will default to `:string`.
:::

### Auto-batching

Auto-batching combines consecutive document addition requests into a batch to be processed together. This significantly speeds up the indexing process.

For document addition requests to be added to the same batch, they need to:

- Target the same index
- Have the same update method (i.e., [POST](/reference/api/documents.md#add-or-replace-documents) or [PUT](/reference/api/documents.md#add-or-update-documents))
- Be immediately consecutive

> Tasks within the same batch share the same values for `startedAt`, `finishedAt`, `duration`, and the same `error` object, if an error occurs.

Reviewer 3: > [...] and the same `error` object, if an error occurs.

This statement is partially wrong, the same error will be thrown for all the tasks if it is due to an "internal" error but if a document is invalid e.g. wrong document id, wrong `_geo` field, therefore only the concerned batch will be erroneous and not processed in the batch, it will be marked as failed with its own error message.

> If a task fails within a batch, the whole batch does not fail.

Reviewer 3: If a task fail within a batch _due to a user error_, the whole batch does not fail.

Author: By "user error", do you mean, e.g., invalid documents?

Reviewer 3: Indeed, that's what I meant :)

> If a task fails due to an invalid document, it will be removed from the batch. The rest of the batch will still process normally. If an [internal](/reference/api/overview.md#errors) error occurs, the whole batch will fail and all tasks within it will share the same `error` object.

Author: ```suggestion
If a task fails due to an invalid document, it will be removed from the batch. The rest of the batch will still process normally. If an [`internal`](/reference/api/overview.md#errors) error occurs, the whole batch will fail and all tasks within it will share the same `error` object.
```
